sw/otbn/crypto/p256_isoncurve.s

```diff
@@ -84,3 +94,120 @@ p256_isoncurve:
   jal       x1, mul_modp
 
   ret
+
+/**
+ * Check if a provided public key is valid.
+ *
+ * For a given public key (x, y), check that:
+ * - x and y are both fully reduced mod p
+ * - (x, y) is on the P-256 curve.
+ *
+ * Note that, because the point is in affine form, it is not possible that (x,
+ * y) is the point at infinity. In some other forms such as projective
+ * coordinates, we would need to check for this also.
+ *
+ * This routine sets `ok` to false if the check fails and immediately exits the
+ * program. If the check succeeds, `ok` is unmodified.
+ *
+ * @param[in] dmem[x]: Public key x-coordinate.
+ * @param[in] dmem[y]: Public key y-coordinate.
+ * @param[out] dmem[ok]: success/failure of basic checks (32 bits)
+ *
+ * clobbered registers: x2, x3, x19, x20, w0, w2, w19 to w29
+ * clobbered flag groups: FG0
+ */
+p256_check_public_key:
+  /* Init all-zero register. */
+  bn.xor   w31, w31, w31
+
+  /* Load domain parameter p.
+       w29 <= dmem[p256_p] = p */
+  li        x2, 29
+  la        x3, p256_p
+  bn.lid    x2, 0(x3)
+
+  /* Load public key x-coordinate.
+       w2 <= dmem[x] = x */
+  li        x2, 2
+  la        x3, x
+  bn.lid    x2, 0(x3)
+
+  /* Compare x to p.
+       FG0.C <= (x < p) */
+  bn.cmp    w2, w29
+
+  /* Fail if FG0.C is false. */
+  csrrs     x2, FG0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _x_valid
+  jal       x0, p256_invalid_input
+  unimp
+
+  _x_valid:
+
+  /* Load public key y-coordinate.
+       w2 <= dmem[y] = y */
+  li        x2, 2
+  la        x3, y
+  bn.lid    x2, 0(x3)
+
+  /* Compare y to p.
+       FG0.C <= (y < p) */
+  bn.cmp    w2, w29
+
+  /* Fail if FG0.C is false. */
+  csrrs     x2, FG0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _y_valid
+  jal       x0, p256_invalid_input
+  unimp
```

Comment:
Should this jump to `p256_invalid_input` rather than invoking `unimp`?

Reply:
Because the jump above is unconditional and can never return (it's with `x0` instead of `x1`, so the call location is not going on the call stack because `x0` ignores writes, see https://opentitan.org/book/hw/ip/otbn/doc/isa.html#jal), we would only reach this line if an attacker had interfered with the jump and forced us not to take it. In this case I want to fail more dramatically than just "invalid input"; I want to immediately stop execution and lock OTBN by throwing an error.

Reply:
The fact it's an unconditional jump was exactly the reason I was asking and suspected it was a failsafe against fault injection. Your explanation makes sense.